libs/parallelism/algorithms/include/hpx/parallel/util/detail/sender_util.hpp

```diff
@@ -0,0 +1,113 @@
+        // and futures being senders. The first version may be slightly more
+        // efficient since a receiver will be connected directly to the returned
+        // future. The second version uses the blocking versions of the parallel
+        // algorithms, and waits on the future inside the transform until the
+        // algorithm has finished.
+
+        // TODO: task policies are mossing for unseq/par_unseq.
+        auto task_policy = std::move(policy)(hpx::execution::task);
+        return hpx::execution::experimental::let_value(
+            std::forward<Predecessor>(predecessor),
+            bound_algorithm<Tag, decltype(task_policy)>{
+                std::move(task_policy)});
+
+        // return hpx::execution::experimental::transform(
+        //     std::forward<Predecessor>(predecessor),
+        //     bound_algorithm<Tag, ExPolicy>{std::forward<ExPolicy>(policy)});
+    }
+
+    // Helper base class for implementing parallel algorithm DPOs. See
+    // tag_fallback documentation for details. Compared to tag_fallback this
+    // adds two tag_fallback_dispatch overloads that are generic for all
+    // parallel algorithms:
+    //
+    //   1. An overload taking a predecessor sender which sends all arguments
+    //      for the regular parallel algorithm overload, except an execution
+    //      policy; and an execution policy.
+    //   2. An overload taking only an execution policy. This overload returns a
+    //      partially applied parallel algorithm, and needs to be supplied a
+    //      predecessor sender. The partially applied algorithm is callable with
+    //      a predecessor sender: partially_applied_algorithm(predecessor). The
+    //      predecessor can also be supplied  using the operator| overload:
+    //      predecessor | partially_applied_parallel_algorithm.
+    template <typename Tag>
+    struct tag_parallel_algorithm : hpx::functional::tag_fallback<Tag>
+    {
+        template <typename Predecessor, typename ExPolicy,
+            typename Enable = std::enable_if_t<
+                std::conjunction_v<hpx::is_execution_policy<ExPolicy>,
+                    std::negation<detail::is_bound_algorithm<Predecessor>>,
+                    hpx::execution::experimental::is_sender<
+                        std::decay_t<Predecessor>>>>>
+        friend auto tag_fallback_dispatch(
+            Tag, Predecessor&& predecessor, ExPolicy&& policy)
+        {
+            return detail::transform_with_bound_algorithm<Tag>(
+                std::forward<Predecessor>(predecessor),
+                std::forward<ExPolicy>(policy));
+        }
+
+        template <typename ExPolicy,
+            HPX_CONCEPT_REQUIRES_(hpx::is_execution_policy<ExPolicy>::value)>
+        friend auto tag_fallback_dispatch(Tag, ExPolicy&& policy)
+        {
+            return hpx::execution::experimental::detail::partial_algorithm<Tag,
+                ExPolicy>{std::forward<ExPolicy>(policy)};
+        }
+    };
```

Comment:
I'm a bit concerned about putting this into `hpx::detail`, however I admit that I don't have the full understanding of the name lookup implications of doing so. For one, the `tag_dispatch`  friend functions here will be injected into that namespace (this is probably of no consequence). Also, how will the compiler find the overloads? Is it because this type is a base class for the algorithm DPOs?

Reply:
Good point. Yes, because it's a base class. I'm also not sure if this will have unintended consequences. Might be better to move it (although that would require separate implementations in `hpx::` and `hpx::ranges::` if we really don't want additional lookup). Then again, the `tag_fallback` base class already pulls in `hpx::functional` for ADL.